File: 2024/day19/python/solution.py
from pathlib import Path
from functools import lru_cache

input_text = (Path(__file__).parent.parent / "input.txt").read_text().strip()

# Parse input
parts = input_text.split("\n\n")
patterns = tuple(p.strip() for p in parts[0].split(","))
designs = parts[1].strip().split("\n")


def can_form(design: str) -> bool:
    """Check if design can be formed by concatenating patterns."""
    @lru_cache(maxsize=None)
    def dp(pos: int) -> bool:
        if pos == len(design):
            return True
        for pattern in patterns:
            plen = len(pattern)
            if design[pos:pos + plen] == pattern:
                if dp(pos + plen):
                    return True
        return False
    return dp(0)


def part1():
    return sum(1 for d in designs if can_form(d))


def count_ways(design: str) -> int:
    """Count number of ways to form design from patterns."""
    @lru_cache(maxsize=None)
    def dp(pos: int) -> int:
        if pos == len(design):
            return 1
        total = 0
        for pattern in patterns:
            plen = len(pattern)
            if design[pos:pos + plen] == pattern:
                total += dp(pos + plen)
        return total
    return dp(0)


def part2():
    return sum(count_ways(d) for d in designs)


if __name__ == "__main__":
    print(f"Part 1: {part1()}")
    print(f"Part 2: {part2()}")

File: 2024/day19/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Solution {
    private static String[] patterns;

    public static void main(String[] args) throws IOException {
        String content = Files.readString(Path.of("../input.txt")).trim();
        String[] parts = content.split("\n\n");

        // Parse patterns (comma-separated on first line)
        patterns = Arrays.stream(parts[0].split(","))
                         .map(String::trim)
                         .toArray(String[]::new);

        // Parse designs (one per line after blank line)
        String[] designs = parts[1].trim().split("\n");

        // Part 1: Count designs that can be formed
        long possibleCount = 0;
        for (String design : designs) {
            if (canForm(design)) {
                possibleCount++;
            }
        }
        System.out.println("Part 1: " + possibleCount);

        // Part 2: Sum the number of ways each design can be formed
        long totalWays = 0;
        for (String design : designs) {
            totalWays += countWays(design);
        }
        System.out.println("Part 2: " + totalWays);
    }

    /**
     * Check if design can be formed by concatenating patterns.
     * Uses dynamic programming with memoization.
     */
    private static boolean canForm(String design) {
        Map<Integer, Boolean> memo = new HashMap<>();
        return dpCanForm(design, 0, memo);
    }

    private static boolean dpCanForm(String design, int pos, Map<Integer, Boolean> memo) {
        if (pos == design.length()) {
            return true;
        }

        if (memo.containsKey(pos)) {
            return memo.get(pos);
        }

        for (String pattern : patterns) {
            int plen = pattern.length();
            if (pos + plen <= design.length() &&
                design.substring(pos, pos + plen).equals(pattern)) {
                if (dpCanForm(design, pos + plen, memo)) {
                    memo.put(pos, true);
                    return true;
                }
            }
        }

        memo.put(pos, false);
        return false;
    }

    /**
     * Count number of ways to form design from patterns.
     * Uses dynamic programming with memoization.
     */
    private static long countWays(String design) {
        Map<Integer, Long> memo = new HashMap<>();
        return dpCountWays(design, 0, memo);
    }

    private static long dpCountWays(String design, int pos, Map<Integer, Long> memo) {
        if (pos == design.length()) {
            return 1;
        }

        if (memo.containsKey(pos)) {
            return memo.get(pos);
        }

        long total = 0;
        for (String pattern : patterns) {
            int plen = pattern.length();
            if (pos + plen <= design.length() &&
                design.substring(pos, pos + plen).equals(pattern)) {
                total += dpCountWays(design, pos + plen, memo);
            }
        }

        memo.put(pos, total);
        return total;
    }
}
